lib/elixir/lib/task/supervised.ex

```diff
@@ -457,43 +469,41 @@ defmodule Task.Supervised do
       # The parent process is telling us to spawn a new task to process
       # "value". We spawn it and notify the parent about its pid.
       {:spawn, position, value} ->
-        {type, pid} = spawn.(parent_pid, normalize_mfa_with_arg(mfa, value))
-        ref = Process.monitor(pid)
-
-        # Schedule a timeout message to ourselves, unless the timeout was set to :infinity
-        timer_ref =
-          case timeout do
-            :infinity -> nil
-            timeout -> Process.send_after(self(), {:timeout, {monitor_ref, ref}}, timeout)
-          end
+        case spawn.(parent_pid, normalize_mfa_with_arg(mfa, value)) do
+          {:ok, type, pid} ->
+            ref = Process.monitor(pid)
+
+            # Schedule a timeout message to ourselves, unless the timeout was set to :infinity
+            timer_ref =
+              case timeout do
+                :infinity -> nil
+                timeout -> Process.send_after(self(), {:timeout, {monitor_ref, ref}}, timeout)
+              end
 
-        send(parent_pid, {:spawned, {monitor_ref, position}, pid})
+            send(parent_pid, {:spawned, {monitor_ref, position}, pid})
 
-        task_info = %{
-          position: position,
-          type: type,
-          pid: pid,
-          timer_ref: timer_ref,
-          timed_out?: false
-        }
+            running_tasks =
+              Map.put(running_tasks, ref, %{
+                position: position,
+                type: type,
+                pid: pid,
+                timer_ref: timer_ref,
+                timed_out?: false
+              })
 
-        running_tasks = Map.put(running_tasks, ref, task_info)
-        stream_monitor_loop(running_tasks, config)
+            stream_monitor_loop(running_tasks, config)
+
+          {:error, :max_children} ->
+            send(parent_pid, {:max_children, monitor_ref})
+            stream_waiting_for_stop_loop(running_tasks, config)
```

Comment:
You can remove the duplication by adding a `catch all` to this receive, something like:

```
  ...
  other ->
    handle_stop_or_exit(other, running_tasks, config)
    stream_monitor_loop(running_tasks, config)
```

And then you just need to create a `stream_stop_loop` that reuses the same logic.

Reply:
Okay. I am not a huge fan of this solution because 1. it complicates the loop by adding indirection since now you don't see all possible messages and 2. it implies that `handle_stop_or_exit` exits the monitor process. But it works. I pushed a commit, if you like it we can get that in in the meantime and I can try and think of a cleaner solution.